pkg/clients/gcp/globaladdress/globaladdress_test.go

```diff
@@ -0,0 +1,243 @@
+				Address:      &address,
+				AddressType:  &addressType,
+				IPVersion:    &ipVersion,
+				Name:         name,
+				Network:      &network,
+				PrefixLength: &prefixLength,
+				Purpose:      &purpose,
+				Subnetwork:   &subnetwork,
+			},
+			observed: &compute.Address{
+				Address:      address + diff,
+				AddressType:  addressType + diff,
+				IpVersion:    ipVersion + diff,
+				Name:         name,
+				Network:      network + diff,
+				PrefixLength: prefixLength + 1,
+				Purpose:      purpose + diff,
+				Subnetwork:   subnetwork + diff,
+			},
+			want: &v1alpha1.GlobalAddressParameters{
+				Address:      &address,
+				AddressType:  &addressType,
+				IPVersion:    &ipVersion,
+				Name:         name,
+				Network:      &network,
+				PrefixLength: &prefixLength,
+				Purpose:      &purpose,
+				Subnetwork:   &subnetwork,
+			},
+		},
+		"OptionalFieldsUnset": {
+			p: &v1alpha1.GlobalAddressParameters{Name: name},
+			observed: &compute.Address{
+				Address:      address,
+				AddressType:  addressType,
+				IpVersion:    ipVersion,
+				Name:         name,
+				Network:      network,
+				PrefixLength: prefixLength,
+				Purpose:      purpose,
+				Subnetwork:   subnetwork,
+			},
+			want: &v1alpha1.GlobalAddressParameters{
+				Address:      &address,
+				AddressType:  &addressType,
+				IPVersion:    &ipVersion,
+				Name:         name,
+				Network:      &network,
+				PrefixLength: &prefixLength,
+				Purpose:      &purpose,
+				Subnetwork:   &subnetwork,
+			},
+		},
+		"ObservedFieldsUnset": {
+			p:        &v1alpha1.GlobalAddressParameters{Name: name},
+			observed: &compute.Address{Name: name},
+			want:     &v1alpha1.GlobalAddressParameters{Name: name},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			UpdateParameters(tc.p, tc.observed)
+			if diff := cmp.Diff(tc.want, tc.p); diff != "" {
+				t.Errorf("UpdateParameters(...): -want, +got:\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestUpdateStatus(t *testing.T) {
+	timestamp := "coolTime"
+	link := "coolLink"
+	users := []string{"coolUser", "coolerUser"}
+	var id uint64 = 3001
+
+	cases := map[string]struct {
+		s        *v1alpha1.GlobalAddressStatus
+		observed *compute.Address
+		want     *v1alpha1.GlobalAddressStatus
+	}{
+		"Reserving": {
+			s: &v1alpha1.GlobalAddressStatus{},
+			observed: &compute.Address{
+				Status:            StatusReserving,
+				CreationTimestamp: timestamp,
+				Id:                id,
+				SelfLink:          link,
+				Users:             users,
+			},
+			want: &v1alpha1.GlobalAddressStatus{
+				ResourceStatus: runtimev1alpha1.ResourceStatus{
+					ConditionedStatus: runtimev1alpha1.ConditionedStatus{
+						Conditions: []runtimev1alpha1.Condition{runtimev1alpha1.Creating()},
+					},
+				},
+				Status:            StatusReserving,
+				CreationTimestamp: timestamp,
+				ID:                id,
+				SelfLink:          link,
+				Users:             users,
+			},
+		},
+		"Reserved": {
```

Comment:
I've noticed that in practice GlobalAddress resources never transition to state Available; they remain in Creating forever. I can't yet figure out why this is, because I see `.status.status` is `RESERVED`.

Reply:
The issue was with `client.Update` overwriting the impending status update with the status at the API server. The solution was to call `client.Update` before updating the status to match the observed resource.